### DISCUSSION TITLE
Zustand state is not updating or is not reflecting in tests

### DISCUSSION THREAD

**Discussion by aqarain** [pmndrs/zustand#2600] [Q&A] [answered]
```
/* store.tsx */

type AppState = {
  item: string;
  setItem: (item: string) => void;
};

const store = createStore<AppState>()((set) => ({
  item: '',
  setItem: (item: string) => set({ item }),
}));


const createBoundedUseStore = ((store) => (selector) => useStore(store)) as <
  S extends StoreApi<unknown>,
>(
  store: S
) => {
  (): ExtractState<S>;
  <T>(selector: (state: ExtractState<S>) => T): T;
};

type ExtractState<S> = S extends { getState: () => infer X } ? X : never;

export const useAppReactStore = createBoundedUseStore(store);
```

```
/* App.tsx */

const App = () => {
  const { item, setItem } = useAppReactStore();


  return (
    <button onClick={() => setItem('Foo')}> Click me </button>
    <h1>{item || 'placeholder'}</h1>
  )
}
```

```
/* App.test.tsx */

describe('App', () => {
  const setItemMock = jest.fn();

  beforeEach(() => {
    (useAppReactStore as jest.Mock).mockReturnValue({
      item: '',
      setItem: setItemMock,
    });
  });

  it('handles Click me button click', () => {
    render(<App />);
    const btn = screen.getByRole('button', { name: 'Click me' });

    fireEvent.click(btn);

    expect(setItemMock).toHaveBeenCalledTimes(1);     // ✅ PASS
    expect(setItemMock).toHaveBeenCalledWith('Foo');  // ✅ PASS
    const heading = screen.getByRole('heading');
    expect(heading).toHaveTextContent('Foo');         // ❌ FAIL
  });
```

I have Zustand store set up like the above in my React app. 
I used the same code in store.tsx as given in the [Zustand guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md#bounded-usestore-hook-for-vanilla-stores). 

The problem is the state change is not reflected. I guess that is why the last assertion is failing.

How can I make that assertion pass? What am I doing wrong here?

**Comment by dbritto-dev** [comment]
@aqarain btw, I encourage you to use testing-library/user-event. Let's fix your current code, you are missing act to call any action in order to trigger a render

So your code should look like this: 

```
act(() => {
  fireEvent.click(btn)
})
```

Note: you need to import act from testing library

**Reply by aqarain** [reply]
@dbritto-dev it is still the same. The last assertion still fails after wrapping it with `act` 😔

Ps. with `userEvent`, the test run slow so sometimes I prefer `fireEvent` 🙂

**Reply by aqarain** [reply]
Do I need to mock the setter function somehow different or something?
Seems like `setItem` is being called correctly but not updating the `item` value in store

**Reply by dbritto-dev** [reply]
@aqarain with userEvent you need to use await. The other way to test it out would be using waitFor.

```
await waitFor(() => {
  expect(heading).toHaveTextContent('Foo');
})
```

Note: in any case you should implement and async function for your tests

**Reply by dbritto-dev** [reply]
BTW, would you mind sharing me a stackblitz demo?

**Reply by aqarain** [reply]
I tried all the below 3 but none worked. 
1. `act(() => { fireEvent.click(btn) })`
2. `userEvent.click(btn)`
3. `await waitFor(() => { expect(heading).toHaveTextContent('Foo') })`

Stackblitz Demo: https://stackblitz.com/edit/vitejs-vite-xdjblu

**Reply by dbritto-dev** [reply, marked as answer]
@aqarain here you go -> https://stackblitz.com/edit/vitejs-vite-pcch7h?file=src%2F__tests__%2FApp.test.tsx

btw, I encourage you to review our testing guide

**Reply by aqarain** [reply]
<img width="1360" alt="Screenshot 2024-06-12 at 10 06 42" src="https://github.com/pmndrs/zustand/assets/20900424/1919e4a3-8179-43e7-a4d4-115564d37a22">


@dbritto-dev I get an error on the stackblitz. Is it for you too?

**Reply by aqarain** [reply]
Thank you so much @dbritto-dev for your help ❤️ 💯
It works fine in Chrome but in Firefox it gives the above error. I guess that is Ok for me.

Btw, the issue is not wrapping `fireEvent` with `act` and it shouldn't be wrapped in `act` (https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#wrapping-things-in-act-unnecessarily)
The issue was mocking of `setItem` 

This is the [testing guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/testing.md), right? Or there is another one?

**Reply by dbritto-dev** [reply]
@aqarain yeah, we don't need act since we are using userEvent, I forgot to remove it. About testing guide, you got it

**Reply by wilsonfaustino** [reply]
I have the same issue: I can check my state updates and how many times the setter has been called, but my tests fail in check changes in the UI. How can we check it?

**Reply by dbritto-dev** [reply]
@wilsonfaustino you can use `waitFor` if you can share me a minimal repro on stackblitz I'd be happy to help

**Reply by wilsonfaustino** [reply]
@dbritto-dev here it is. https://stackblitz.com/~/github.com/wilsonfaustino/poc-zustand

thanks in advance.

**Reply by dbritto-dev** [reply]
@wilsonfaustino here you go 

To implement `waitFor` to check ui updates

```tsx
// src/components/app/counter.test.tsx
import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import { Counter } from './counter'

describe('Counter', () => {
  it('should render the counter component', async () => {
    render(<Counter />)

    await waitFor(() => {
      expect(screen.getByText('Contador:')).toBeInTheDocument()
      expect(screen.getAllByText(0)[0]).toHaveStyle({ transform: 'none' })
      expect(screen.getAllByText(0)[1]).toHaveStyle({ transform: 'none' })
      expect(screen.getAllByText(0)[2]).toHaveStyle({ transform: 'none' })
    }, { timeout: 1500 })
  })

  it('should increment the counter when the increment button is clicked', async () => {
    render(<Counter />)

    const incrementButton = screen.getByText('Incrementar')

    fireEvent.click(incrementButton)

    await waitFor(() => {
      expect(screen.getAllByText(1)[2]).toHaveStyle({ transform: 'none' })
    }, { timeout: 1500 })
  })

  it('should decrement the counter when the decrement button is clicked', async () => {
    render(<Counter />)

    const decrementButton = screen.getByText('Decrementar')

    fireEvent.click(decrementButton)

    await waitFor(() => {
      expect(screen.getAllByText(9)[0]).toHaveStyle({ transform: 'none' })
      expect(screen.getAllByText(9)[1]).toHaveStyle({ transform: 'none' })
      expect(screen.getAllByText(9)[2]).toHaveStyle({ transform: 'none' })
    }, { timeout: 1500 })
  })
})
```

To set `aria-hidden` on elements that are not visible

```tsx
// src/components/ui/counter-display.tsx
const Digit = ({ place, value }: { place: number; value: number }) => {
  const valueRoundedToPlace = Math.floor(value / place)
  const animatedValue = useSpring(valueRoundedToPlace)

  useEffect(() => {
    animatedValue.set(valueRoundedToPlace)
  }, [animatedValue, valueRoundedToPlace])

  return (
    <div style={{ height }} className="relative w-[1ch] tabular-nums">
      {[...Array(10).keys()].map((i) => (
        <Number key={i} mv={animatedValue} number={i} />
      ))}
    </div>
  )
}

const Number = ({ mv, number }: { mv: MotionValue; number: number }) => {
  const y = useTransform(mv, (latest) => {
    const placeValue = latest % 10
    const offset = (10 + number - placeValue) % 10

    let memo = offset * height

    if (offset > 5) {
      memo -= 10 * height
    }

    return memo
  })

  return (
    <motion.span style={{ y }} aria-hidden={y.get() === 0 ? 'false' : 'true'} className="absolute inset-0 flex items-center justify-center">
      {number}
    </motion.span>
  )
}
```
To make jest and jest-dom types available

```json
// tsconfig.app.json
{
  "compilerOptions": {
    "composite": true,
    "tsBuildInfoFile": "./node_modules/.tmp/tsconfig.app.tsbuildinfo",
    "target": "ES2020",
    "useDefineForClassFields": true,
    "lib": ["ES2020", "DOM", "DOM.Iterable"],
    "module": "ESNext",
    "skipLibCheck": true,

    /* Bundler mode */
    "moduleResolution": "bundler",
    "allowImportingTsExtensions": true,
    "resolveJsonModule": true,
    "isolatedModules": false,
    "moduleDetection": "force",
    "noEmit": true,
    "jsx": "react-jsx",

    /* Linting */
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noFallthroughCasesInSwitch": true,
    
    /* Paths for Shadcn UI */
    "baseUrl": ".",
    "paths": {
      "@/*": [
        "./src/*"
      ]
    },
    "types": ["jest", "@testing-library/jest-dom"]
  },
  "include": ["src", "./setup-tests.ts"]
}
```

**Reply by wilsonfaustino** [reply]
I will give it a try as soon as I leave work. Thank you @dbritto-dev <3

**Reply by dbritto-dev** [reply]
@wilsonfaustino lmk if you have any questions, btw in this case is a bit tricky to test ui changes due to many elements in the dom that are available but no visible. The only way to check if they are visible is through their styles and you can check that with `.toHaveStyle()`

**Reply by wilsonfaustino** [reply]
@dbritto-dev thank you very much for the answer. I need to foccus a little more in tests.
